src/Types.jl

```diff
@@ -424,13 +424,27 @@ function write_env_usage(manifest_file::AbstractString)
     close(io)
 end
 
+project_toplevel_spec = [
+    "desc", "keywords", "license", "name", "authors", "uuid",
+    "deps", "extras", "targets",
+]
+
+function enforce_project!(project)
+    for key in keys(project)
+        key in project_toplevel_spec ||
+            cmderror("Unrecognized entry in project file: '$key'")
```

Comment:
Should this be enforced by Pkg? What if I want to add some other metadata field to my project file?

Reply:
Good point. Maybe warn and remove from dictionary?

Reply:
I would prefer input checks be as strong as possible. Maybe a toplevel toml field that users can shove anything they want into.

Reply:
What if we add another allowed field in the future? Then a newer project file would error on older Pkg versions.

Reply:
Another good point! If we want really strict input that would mean we would have to add a pkg-version field and map a pkg-version with some sort of spec for toml files. ...but that seems overzealous at this point :laughing: .

For now, how about: a toplevel `user`(?) field + warn about any unrecognized fields. If you feel really strongly against this check, then we can make no warnings the default allow users to opt-in with some '--strict-parse' flag.

Reply:
It’s fine for the package manager to be strict about the structure of things it knows about but it should just ignore anything that it doesn’t know about. This is really crucial for adding features in the future and continuing to work gracefully with older versions of the package manager and language in general. Ideally it should only verify that the structure confirms to what it requires to function correctly.

Reply:
> This is really crucial for adding features in the future and continuing to work gracefully with older versions of the package manager and language in general.

If it ignores everything except what we have implemented today, then adding features will break existing file that uses the name of the feature since everything is in the same namespace. Enforcing that all user customization need to happen in its own namespace `metadata` solves that.